ingestion/src/kvkafka.rs

```diff
@@ -229,27 +234,35 @@ async fn read_kafka(group_name: String, tx: mpsc::Sender<Msg>) {
 
     // Consume the kafka queue infinitely
     loop {
-        // https://docs.rs/kafka/latest/src/kafka/consumer/mod.rs.html#155
-        // poll asks for next available chunk of data as a MessageSet
-        match consumer.poll() {
-            Ok(sets) => {
-                for msgset in sets.iter() {
-                    for msg in msgset.messages() {
-                        if let Err(e) = parse_message(msg.value, &tx).await {
-                            eprintln!("{}", e);
+        tokio::select! {
+            _ = cancel_token.cancelled() => {
+                eprintln!("cancellation token triggered");
+                break;
+            }
+            // https://docs.rs/kafka/latest/src/kafka/consumer/mod.rs.html#155
+            // poll asks for next available chunk of data as a MessageSet
+            poll_result = async { consumer.poll() } => {
+                match poll_result {
+                    Ok(sets) => {
+                        for msgset in sets.iter() {
+                            for msg in msgset.messages() {
+                                if let Err(e) = parse_message(msg.value, &tx).await {
+                                    eprintln!("{}", e);
+                                }
+                            }
+                            if let Err(e) = consumer.consume_messageset(msgset) {
+                                eprintln!("{}", e);
+                            }
```

Comment:
Another option with cancellation tokens (which I've never used before by the way) is to skip the `select` altogether.
I guess it's a matter of preference and it doesn't do exactly what the select code does (in particular, it doesn't play well with the `Err` branch of the match statement), but I find it easier to reason about.

```suggestion
        if cancel_token.is_cancelled() {
            eprintln!("cancellation token triggered");
            break;
        }

        // https://docs.rs/kafka/latest/src/kafka/consumer/mod.rs.html#155
        // poll asks for next available chunk of data as a MessageSet
        match consumer.poll() {
            // process message sets
            ...
        }
```

Reply:
Yes, I forgot that our kafka library is sync. In that case I think this is equivalent to the select, and probably the way to go. It has another problem though: In the pathological case where the kafka queue is empty when we cancel, This will hang indefinitely (or at least the 90 seconds until systemd sends `SIGKILL`...). It's also a problem that we're doing blocking IO on a non-blocking task.

We can find a way around this, but I'm starting to think it would make more sense to switch to `rdkafka` which supports async.

Reply:
Why indefinitely? If the queue is empty I'd expect `consumer.poll()` to return an error (?), after which we wait 5 seconds and check if the token was cancelled before polling again.
But maybe it's not so simple and I agree we should probably switch to `rdkafka`.

> It's also a problem that we're doing blocking IO on a non-blocking task.

This is the bug you mentioned in the other comment? That we are simply using `spawn` instead of `spawn_blocking`?

Reply:
> Nope. What I was talking about is that at the moment we commit offsets without waiting for the associated DB inserts to complete. Related to that the DB inserts being on a separate task means they aren't covered by the graceful shutdown, but a solution to the first problem with probably also solve the second.

I think it would be covered indirectly by the sender being dropped? But oof. good catch, it probably makes more sense to parallelize over message sets instead of single messages.